Merging DependencyCollector.Tests and DependencyCollector.Shared.Tests

eddynaka opened the pull request:
Refs #1769

## Changes
Merging DependencyCollector.Tests and DependencyCollector.Shared.Tests
Adding netcoreapp3.1 to TargetFrameworks

### Checklist
- [X] I ran Unit Tests locally.
- [ ] CHANGELOG.md updated with one line description of the fix, and a link to the original issue if available.

For significant contributions please make sure you have completed the following items:

- [ ] Design discussion issue #
- [ ] Changes in public surface reviewed

The PR will trigger build, unit tests, and functional tests automatically. Please follow [these](https://github.com/Microsoft/ApplicationInsights-dotnet/blob/develop/.github/CONTRIBUTING.md) instructions to build and test locally.

### Notes for authors:
- FxCop and other analyzers will fail the build. To see these errors yourself, compile localy using the Release configuration.

### Notes for reviewers:

- We support [comment build triggers](https://docs.microsoft.com/azure/devops/pipelines/repos/github?view=azure-devops&tabs=yaml#comment-triggers)
  - `/AzurePipelines run` will queue all builds
  - `/AzurePipelines run <pipeline-name>` will queue a specific build

eddynaka commented:
@cijothomas / @TimothyMothra , the following change has some modifications:

Before this change, we had:
- NET45: 222 tests, which 28 (dont run) / 191 (passed) / 3 (skipped) => **222** Total
- NETCoreApp2.1: 202 (passed) / 1 (skipped) => 203 Total

After this change, we have:
- NET45: 222 tests, which 28 (dont run) / 360 (passed) / 3 (skipped) / **6 (failed)** => **397** Total
- NETCoreApp2.1: 202 (passed) / 1 (skipped) => 203 Total
- NETCoreApp3.1: 201 (passed) / 1 (skipped) / **1 (error)** => 203 Total

Debugging the failling tests, we have some changes from values that I don't know if that is expected, for example, for netcore3.1, we are missing some headers. For net45, we have different values.

TimothyMothra commented:
Overall this looks good.
I'm curious about the changes to `ObjectInstanceBasedOperationHolder` in some of the tests.
And I see that some tests are failing that we need time to investigate.

eddynaka commented:
@cijothomas / @TimothyMothra , we have the following situation:

DependencyTrackingTelemetryModuleTestNetCore
- TestDependencyCollectionWithW3COnRequestIdOff: for netcoreapp3.1, this isn't working. Missing header. Debugging, it went to version V3, which has a complete different logic from V2 (netcoreapp2.1)

ProfilerHttpProcessingTest (net45 only): when i moved to the test project, **35** tests were added (the file existed in the shared project, but it wasn't running). Below, the tests that doesn't work:
- RddTestHttpProcessingProfilerOnBeginAddsLegacyHeadersAreEnabled
- RddTestHttpProcessingProfilerOnBeginOnEnd

DesktopDiagnosticSourceHttpProcessingTests (net45 only): when i moved to the test project, **11** tests were added (the file existed in the shared project, but it wasn't running). Below, the tests that doesn't work:
- RddTestHttpDesktopProcessingFrameworkOnBeginAddsLegacyHeaders

FrameworkSqlProcessingTest (net45 only): when i moved to the test project, **7** tests were added (the file existed in the shared project, but it wasn't running). Below, the tests that doesn't work:
- RddTestSqlProcessingFrameworkSendsCorrectTelemetryStoredProc

TimothyMothra commented:
Regarding these test classes:
- ProfilerHttpProcessingTest 
- DesktopDiagnosticSourceHttpProcessingTests 
- RddTestHttpDesktopProcessingFrameworkOnBeginAddsLegacyHeaders

Eddy said that these tests weren't running in the Develop branch. 
I think we should either remove them or mark them as `[Ignore("Test class out of date")].`
Give me time to verify, before making this change.

Regarding this:
- DependencyTrackingTelemetryModuleTestNetCore

It sounds like adding netcoreapp3.1 changed something in the execution.
Eddy, if you can debug this test you might find a code branch in the Product code that is being skipped. 
For example, if they use (#IF NETSTANDARD1.X OR NETSTARDARD2.0) this would skip a NETSTANDARD2.1 (aka NetCore 3.x)

TimothyMothra commented:
> Regarding these test classes:
> 
> * ProfilerHttpProcessingTest
> * DesktopDiagnosticSourceHttpProcessingTests
> * FrameworkSqlProcessingTest
> 
> Eddy said that these tests weren't running in the Develop branch.
> I think we should either remove them or mark them as `[Ignore("Test class out of date")].`
> Give me time to verify, before making this change.


I verified that these tests weren't being run in DEVELOP.
Eddy, please open a new Github Issue to investigate these out of date tests.
Then mark the TestClass with this attribute: `[Ignore("Test class out of date. Github Issue #####")]`. Including the issue Number that you created.

TimothyMothra commented:
**Remaining task:**

class: DependencyTrackingTelemetryModuleTestNetCore
test: TestDependencyCollectionWithW3COnRequestIdOff

netcoreapp2.1 - pass
netcoreapp3.1 - fail

eddynaka commented:
Debugging the failing test, I found some differences between Netcoreapp2.1 and Netcoreapp3.1:
- Netcoreapp2.1: go as HttpInstrumentationVersion.V2
- Netcoreapp3.1: go as HttpInstrumentationVersion.V3

In the class, we have this: [HttpCoreDiagnosticSourceListener.cs](https://github.com/microsoft/ApplicationInsights-dotnet/blob/develop/WEB/Src/DependencyCollector/DependencyCollector/HttpCoreDiagnosticSourceListener.cs#L544), but, yet, I couldn't find where it fills the property Request-Id in Netcoreapp2.1.

TimothyMothra commented on Everything.sln:
```diff
@@ -91,8 +91,6 @@ Project("{D954291E-2A0B-460D-934E-DC6B0785DB48}") = "TestFramework.Shared.Base",
 Project("{D954291E-2A0B-460D-934E-DC6B0785DB48}") = "TestFramework.Shared.Web", "WEB\Src\TestFramework\Shared\TestFramework.Shared.Web.shproj", "{9718F051-147F-4F5F-9FF3-C926430EFCF7}"
 EndProject
-Project("{D954291E-2A0B-460D-934E-DC6B0785DB48}") = "DependencyCollector.Shared.Tests", "WEB\Src\DependencyCollector\Shared.Tests\DependencyCollector.Shared.Tests.shproj", "{ACE58393-3419-4FCA-87CC-C33EB756C7E4}"
-EndProject
```
Need to also remove this from the Web.sln

TimothyMothra commented:
Two things noteworthy:

This test is NOT expecting a RequestId:
https://github.com/microsoft/ApplicationInsights-dotnet/blob/96d8f7011e372da703ec141d9f106a0bff28b714/WEB/Src/DependencyCollector/DependencyCollector.Tests/DependencyTrackingTelemetryModuleTestNetCore.cs#L356

BUT, the test is expecting NetCore2 to inject a RequestIdHeader.
https://github.com/microsoft/ApplicationInsights-dotnet/blob/96d8f7011e372da703ec141d9f106a0bff28b714/WEB/Src/DependencyCollector/DependencyCollector.Tests/DependencyTrackingTelemetryModuleTestNetCore.cs#L485-L499

This test fails because NetCore3 is not injecting the same headers.
I added the `#elif NETCOREAPP3_1` to force this test to pass, but we need to confirm what is the expected behavior here.
@cijothomas do you have any context on NetCore and Request-Id headers?

cijothomas commented:
@TimothyMothra There are differences in correlationin .net 3 and 2. 
To unblock - proceed with what you did - conditionallyu test for .netcoreapp3.1.
And open an issue for this, you can assign to me.
